src/mbgl/style/sources/tile_source.cpp

```diff
@@ -0,0 +1,108 @@
+                       uint16_t tileSize,
+                       SourceType sourceType)
+    : Source(makeMutable<Impl>(sourceType, std::move(id), tileSize)),
+      urlOrTileset(std::move(urlOrTileset_)) {}
+
+TileSource::~TileSource() = default;
+
+const TileSource::Impl& TileSource::impl() const {
+    return static_cast<const Impl&>(*baseImpl);
+}
+
+const variant<std::string, Tileset>& TileSource::getURLOrTileset() const {
+    return urlOrTileset;
+}
+
+std::optional<std::string> TileSource::getURL() const {
+    if (urlOrTileset.is<Tileset>()) {
+        return {};
+    }
+
+    return urlOrTileset.get<std::string>();
+}
+
+uint16_t TileSource::getTileSize() const {
+    return impl().getTileSize();
+}
+
+void TileSource::loadDescription(FileSource& fileSource) {
+    if (urlOrTileset.is<Tileset>()) {
+        baseImpl = makeMutable<Impl>(impl(), urlOrTileset.get<Tileset>());
+        loaded = true;
+        observer->onSourceLoaded(*this);
+        return;
+    }
+
+    if (req) {
+        return;
+    }
+
+    const auto& rawURL = urlOrTileset.get<std::string>();
+    const auto& url = util::mapbox::canonicalizeSourceURL(fileSource.getResourceOptions().tileServerOptions(), rawURL);
+
+    req = fileSource.request(Resource::source(url), [this, url, &fileSource](const Response& res) {
+        if (res.error) {
+            observer->onSourceError(*this, std::make_exception_ptr(std::runtime_error(res.error->message)));
+        } else if (res.notModified) {
+            return;
+        } else if (res.noContent) {
+            observer->onSourceError(*this, std::make_exception_ptr(std::runtime_error("unexpectedly empty TileJSON")));
+        } else {
+            conversion::Error error;
+            std::optional<Tileset> tileset = conversion::convertJSON<Tileset>(*res.data, error);
+            if (!tileset) {
+                observer->onSourceError(*this, std::make_exception_ptr(util::StyleParseException(error.message)));
+                return;
+            }
+            const auto& tileServerOptions = fileSource.getResourceOptions().tileServerOptions();
+            if (tileServerOptions.uriSchemeAlias() == "mapbox") {
+                util::mapbox::canonicalizeTileset(tileServerOptions, *tileset, url, getType(), getTileSize());
+            }
+
+            this->setTilesetOverrides(*tileset);
+
+            bool changed = impl().tileset != *tileset;
+
+            baseImpl = makeMutable<Impl>(impl(), *tileset);
+            loaded = true;
+
+            observer->onSourceLoaded(*this);
+
+            if (changed) {
+                observer->onSourceChanged(*this);
+            }
+        }
+    });
+}
+
+void TileSource::setTilesetOverrides(Tileset&) {
+    // Default empty implementation, subclasses can override
+}
+
+bool TileSource::supportsLayerType(const mbgl::style::LayerTypeInfo* info) const {
+    return mbgl::underlying_type(Tile::Kind::Raster) == mbgl::underlying_type(info->tileKind);
```

Comment:
I think we need to force derived classes to implement this. It does not make sense to provide a default implementation. 

You can remove it from the `TileSource` header and implementation, because it is already defined in `Source`. 

This will make `TileSet` an abstract class (which is OK, because it is not being instantiated directly anywhere).

Reply:
Sounds good!